### DISCUSSION TITLE
GraalVM native build not working with logback

### DISCUSSION THREAD

**Discussion by Bazalii** [quarkusio/quarkus#30342] [Q&A]
I am trying to build a native executable and native image using graalVM and Jib. I use ./gradlew build command. Here are the properties that are in the application.properties:

```
quarkus.container-image.build=true
quarkus.container-image.builder=docker
quarkus.native.container-build=true
quarkus.native.container-runtime=docker
quarkus.package.type=native
```

During build I get such exception: 

```
UnresolvedElementException: Discovered unresolved type during parsing: sun.reflect.Reflection.
This error is reported at image build time because class ch.qos.logback.classic.spi.PackagingDataCalculator 
is registered for linking at image build time by command line
```

I tried adding that in build.gradle:

```
graalvmNative {
    binaries {
        all {
            resources.autodetect()
        }

        named("main") {
            imageName.set("application")

            buildArgs.add("--verbose")
            buildArgs.add("--add-opens=java.base/java.nio=ALL-UNNAMED")
            buildArgs.add("--add-opens=java.base/jdk.internal.misc=ALL-UNNAMED")
            buildArgs.add("--add-opens=java.base/jdk.internal.ref=ALL-UNNAMED")
            buildArgs.add("--trace-class-initialization=ch.qos.logback.classic.Logger")
            buildArgs.add("--initialize-at-build-time=ch.qos.logback")
        }
    }
}

nativeBuild {
    buildArgs("-H:ReflectionConfigurationFiles=${file("reflection-config.json").path}")
}
```

and added this dependency to plugins in build.gradle:

`id("org.graalvm.buildtools.native") version "0.9.18"`

Unfortunately, that did not work..

Is there an option to use graalVM native build with logback?

[a comment by an automated account is not shown]

**Comment by geoand** [comment]
We don't support logback so therefore you mileage with in native mode will vary

**Reply by Bazalii** [reply]
So, there are no ways to solve this problem, isn't it?

**Reply by geoand** [reply]
I can' t say, as we have not tried to use logback

**Reply by fbuetler** [reply]
@geoand just being curious, why does quarkus not support logback? Is it because of the native build or was there a different reason for this design decision? Asking because we are experiencing exactly the same issue with graalvm and logback.

**Reply by geoand** [reply]
When Quarkus started back in 2018, Logback did not work at all in native. Things may or may not have changed since then

**Reply by fbuetler** [reply]
Thanks for the quick clarification 👍🏻

**Reply by geoand** [reply]
🙏🏼

**Reply by Manfred73** [reply]
Would be great if logback was supported.

We have several components in Spring and several in Quarkus for a project. Also some shared libraries.
We started with Spring using logback and logstash for structured json logging.
For some new components we started using Quarkus native because of smaller footprint (Spring native was still in beta).
We finally ended up having two separate shared libraries for structured logging: one for Quarkus making use of quarkus-logging-json and one for spring making use of logback and logstash. And we make sure that no other shared libraries that we want to share between Quarkus and Spring components make use of logback. Not ideal to maintain two different libraries which are framework dependent, but for now this works for us.
Ideally I think one would stick to one framework within a project.

**Reply by geoand** [reply]
What exactly from logback do you need?

**Reply by Manfred73** [reply]
We wanted to use it for structured logging (same way as we were using it for spring components). 
We had some shared libraries which we also used in Quarkus components, but the problem when they are transitively included via other dependencies is that Quarkus native build will fail with the errors I mentioned in my original message. So we switched to quarkus-logging-json and excluded all logback libraries from dependencies that provided them.

**Comment by Manfred73** [comment]
I'm running into the same issue, using some external libraries that use logback.
So I tried to follow these instructions:

- https://stackoverflow.com/questions/72770461/graalvm-native-image-can-not-compile-logback-dependencies
- https://github.com/quarkusio/quarkus/issues/26613

I'm now using a `ReflectionsConfig` class listing all the logback classes mentioned in the first link (and also added the class `PackagingDataCalculator` it's complaining about).

```
import ch.qos.logback.classic.AsyncAppender;
import ch.qos.logback.classic.encoder.PatternLayoutEncoder;
import ch.qos.logback.classic.pattern.DateConverter;
import ch.qos.logback.classic.pattern.LevelConverter;
import ch.qos.logback.classic.pattern.LineSeparatorConverter;
import ch.qos.logback.classic.pattern.LoggerConverter;
import ch.qos.logback.classic.pattern.MessageConverter;
import ch.qos.logback.classic.pattern.ThreadConverter;
import ch.qos.logback.classic.spi.PackagingDataCalculator;
import ch.qos.logback.core.ConsoleAppender;
import ch.qos.logback.core.FileAppender;
import io.quarkus.runtime.annotations.RegisterForReflection;

@RegisterForReflection(targets = {
		PackagingDataCalculator.class,  // native build complains about this one
		AsyncAppender.class,
		PatternLayoutEncoder.class,
		DateConverter.class,
		LevelConverter.class,
		LineSeparatorConverter.class,
		LoggerConverter.class,
		MessageConverter.class,
		ThreadConverter.class,
		ConsoleAppender.class,
		FileAppender.class
})
public class ReflectionsConfig {
}
```

I've updated the native profile in the `pom.xml` with:

`<quarkus.native.additional-build-args>--initialize-at-build-time=ch.qos.logback</quarkus.native.additional-build-args>`.

Unfortunately, this doesn't solve the issue.
Is there a way to get this working when using 3rd party / external libraries that use logback?

**Comment by salespaulo** [comment]
You can use this https://www.graalvm.org/latest/reference-manual/native-image/guides/configure-with-tracing-agent/ to generate configuration needs to compile reflections and more...
